src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -20,8 +20,64 @@ internal enum NullableAnnotation : byte
         Nullable,     // Explicitly set by flow analysis
     }
 
+    /// <summary>
+    /// The nullable state of an rvalue computed in <see cref="NullableWalker"/>.
+    /// When in doubt we conservatively use <see cref="NullableFlowState.NotNull"/>
+    /// to minimize diagnostics.
+    /// </summary>
+    internal enum NullableFlowState : byte
+    {
+        NotNull,
+        MaybeNull
+    }
+
+    /// <summary>
+    /// A type and its corresponding flow state resulting from evaluating an rvalue expression.
+    /// </summary>
+    [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
+    internal struct TypeWithState
+    {
+        public TypeSymbol Type { get; }
+        public NullableFlowState State { get; }
+        public bool HasNullType => Type is null;
+        public bool MaybeNull => State == NullableFlowState.MaybeNull;
+        public bool NotNull => State == NullableFlowState.NotNull;
+        public bool IsDefault => Type is null;
+        public static TypeWithState ForType(TypeSymbol type) => new TypeWithState(type, type?.CanContainNull() == true ? NullableFlowState.MaybeNull : NullableFlowState.NotNull);
+        public TypeWithState(TypeSymbol type, NullableFlowState state) => (Type, State) = (type, state);
+        public void Deconstruct(out TypeSymbol type, out NullableFlowState state) => (type, state) = (Type, State);
+        public static implicit operator TypeWithState((TypeSymbol type, NullableFlowState state) a) => new TypeWithState(a.type, a.state);
+        public string GetDebuggerDisplay() => $"{{Type:{Type?.GetDebuggerDisplay()}, State:{State}{"}"}";
+        public TypeWithState WithTopLevelNonNullability() => new TypeWithState(Type, NullableFlowState.NotNull);
+        public TypeSymbolWithAnnotations ToTypeSymbolWithAnnotations()
+        {
+            NullableAnnotation annotation;
+            switch (this.State)
+            {
+                case NullableFlowState.NotNull:
+                    annotation = NullableAnnotation.NotNullable;
+                    break;
+                default: // MaybeNull
+                    annotation = NullableAnnotation.Nullable;
+                    break;
+            }
+
+            return TypeSymbolWithAnnotations.Create(this.Type, annotation);
+        }
+    }
+
     internal static class NullableAnnotationExtensions
     {
+        public static NullableFlowState ToState(this NullableAnnotation annotation)
```

Comment:
>ToState [](start = 40, length = 7)

nit: I'd be tempted to inline this method, as it is a bit strange: It feels like you should not be able to go from an annotation to a state without also looking at the type.